“增加单测实现除静态断言之外的全覆盖测试，规范化README.md，增加注释”

Simon25772 opened the pull request:
“增加单测实现除静态断言之外的全覆盖测试，规范化README.md，增加注释”

Commit:
test: add tests

Simon25772 commented:
忘记设置为Draft了，稍后解决

YdrMaster commented on README.md:
```diff
@@ -12,4 +12,55 @@
 ![GitHub commit activity](https://img.shields.io/github/commit-activity/m/InfiniTensor/ndarray-layout)
 
-This crate provides definitions and transformations for multi-dimensional array data layouts.
+
```
这个文件格式需要规范一下

Simon25772 replied:
好滴

YdrMaster commented on src/fmt.rs:
```diff
@@ -1,72 +1,119 @@
-﻿use crate::ArrayLayout;
+﻿// 引入 crate 中的 ArrayLayout 结构体
```
use 就不用加注释了😂

YdrMaster commented on src/fmt.rs:
```diff
@@ -1,72 +1,119 @@
+    /// 调用者需要确保指针 `ptr` 有效，并且指向的内存区域包含足够的元素，
+    /// 同时偏移量计算不会导致越界访问。
+    ///
+    /// # 参数
```
RustDoc 也是 Markdown，注意规范，标题和正文之间空行

YdrMaster commented on src/fmt.rs:
```diff
@@ -1,72 +1,119 @@
+    /// - `f`: 用于格式化输出的 `fmt::Formatter` 引用。
+    /// - `ptr`: 指向要格式化的数组元素的裸指针。
+    ///
+    /// # 返回值
```
一般 RustDoc 不用写这种套话

YdrMaster commented:
用力过猛😂不要每行都注释啊

Simon25772 commented:
> 用力过猛😂不要每行都注释啊

好滴

Commit:
style(README): 规范格式

Commit:
style: 解决之前注释不规范的问题

Commit:
style: 格式化代码

Commit:
style: 使用vscode自动格式化代码

Commit:
style: 格式化Rust Document

YdrMaster commented on src/lib.rs:
```diff
@@ -1,4 +1,4 @@
-#![doc = include_str!("../README.md")]
+# ![doc = include_str!("../README.md")]
```
多加了一个空格

YdrMaster commented on src/lib.rs:
```diff
@@ -40,7 +40,9 @@ impl<const N: usize> Drop for ArrayLayout<N> {
 }
 
 /// 元信息存储顺序。
+
```
多了空行，下面还有就不每个都写了

YdrMaster commented on src/lib.rs:
````diff
@@ -57,7 +59,7 @@ impl<const N: usize> ArrayLayout<N> {
     /// assert_eq!(layout.offset(), 20);
     /// assert_eq!(layout.shape(), &[2, 3, 4]);
     /// assert_eq!(layout.strides(), &[12, -4, 1]);
-    /// ```
````
这个空格不能去掉

YdrMaster commented on src/transform/merge.rs:
```diff
@@ -76,7 +76,8 @@ impl<const N: usize> ArrayLayout<N> {
         let strides = content.strides();
 
-        let merged = args.iter().map(|arg| arg.len).sum::<usize>();
+        // 修改BUG
```
“修改 Bug” 是你这次提交的目的，不是这个代码的目的，怎么会作为注释呢😂。

这个改动的目的是：`// 兼容无意义的输入：合并 0 或 1 个维度不会改变布局`。

所以，

```rust
if len == 0 {
    continue;
}
```

是不是应该改成：

```rust
if len < 2 {
    continue;
}
```

?

另外干脆把这个改成 fold 吧，顺便带一个 bool，如果所有 len 都是 0 或 1 直接 `return self` 了。

YdrMaster commented on src/transform/merge.rs:
```diff
@@ -109,6 +110,10 @@ impl<const N: usize> ArrayLayout<N> {
                 }
             }
+
+            // 修改BUG
```
同上，这里去掉注释好了

YdrMaster commented on README.md:
```diff
@@ -12,4 +12,55 @@
+
+### 多维数组布局管理
+
+* ArrayLayout 结构体支持指定任意维度的数组布局，通过 new 方法可以创建具有指定形状、步长和偏移量的布局。
```
无序列表写成

```markdown
- 无序；
- 列表；
```

这种，标号中划线，结尾分号。`*` 和 `-` 对于 markdown 没区别但是我习惯 `-`。

Simon25772 committed:
ci: 更新action脚本

Commit:
style: 更改格式

Commit:
docs(README): 调整

Commit:
style(README,lib,merge): 调整

Commit:
style(lib,merge): 调整

Commit:
docs(README): add the coverage tag, but the link uses my own codecov

YdrMaster commented on README.md:
```diff
@@ -12,4 +13,43 @@
 ![GitHub commit activity](https://img.shields.io/github/commit-activity/m/InfiniTensor/ndarray-layout)
 
-This crate provides definitions and transformations for multi-dimensional array data layouts.
+ndarray-layout 是一个用于处理多维数组布局的 Rust 库，它提供了 ArrayLayout 结构体，用于高效管理和操作多维数组的元信息，如形状、步长和偏移量等。这个库在处理多维数组时，提供了灵活且高效的布局管理方式，能够满足不同场景下对数组布局的操作需求。
```
可以再提高一下，最规范的写法是

- 用 ` 包围所有代码中摘录的东西，例如类型和方法的名字；
- 用 * 斜体包围所有其他名字比如项目本身；

YdrMaster commented on src/fmt.rs:
```diff
@@ -89,6 +89,7 @@ impl<const N: usize> ArrayLayout<N> {
 }
 
 #[test]
+
```
还是有一些多空行，在 github 上自己看一遍比较清楚

YdrMaster commented on src/lib.rs:
```diff
@@ -214,6 +223,7 @@ impl<const N: usize> ArrayLayout<N> {
     #[inline]
     fn ptr_allocated(&self) -> Option<NonNull<usize>> {
         const { assert!(N > 0) }
+        // ndim>N则content是ptr,否则是元组
```
一般注释也要注意格式，养成好习惯，尤其这里有英文逗号

YdrMaster commented on src/transform/merge.rs:
```diff
@@ -76,7 +76,13 @@ impl<const N: usize> ArrayLayout<N> {
         let strides = content.strides();
 
-        let merged = args.iter().map(|arg| arg.len).sum::<usize>();
+        let (merged, flag) = args.iter().fold((0, true), |(acc, _f), arg| {
```
还有个更简单的判断方法，merged 如果等于 args.len() 就说明没 merge 了

YdrMaster replied:
我刚想到的😂

Commit:
style: remove redundant blank lines

Commit:
docs: improve the readability of README.md

Commit:
style: standardized annotation format

Commit:
refactor(transform/merge): simplify the merging judgment conditions

The original judgment logic for accumulating flags in fold was changed to compare merged with `args.len()` to determine whether to merge. This judgment is equivalent but more concise.

Commit:
docs: append Simon25772 to authors list

Commit:
docs(CHANGELOG): update CHANGELOG

Commit:
style(transform/broadcast): remove redundant blank lines

Commit:
style(transform/index): remove redundant blank lines

Commit:
docs(transform/broadcast): 更正注释
